res/controllers/PIONEER-DDJ-ERGO-scripts.js

```diff
@@ -0,0 +1,441 @@
+    const deck = status-0xB4;
+    const effect = selectedEffects[deck];
+    //if called with MSB
+    if (control < 0x20) {
+        //if fx selector turned enough
+        if (control === 0x00 && previousEffectKnobPosition !== value/5) {
+            //switch to next or previous fx in list
+            if (previousEffectKnobPosition > value/5) {
+                engine.setValue(effectsStrings[deck][effect], "effect_selector", -1);
+            } else {
+                engine.setValue(effectsStrings[deck][effect], "effect_selector", 1);
+            }
+            previousEffectKnobPosition = value/5;
+        }
+        //if fx knob 1 turned
+        if (control === 0x02) {
+            engine.setValue(effectsStrings[deck][effect], "meta", value/127);
+        }
+    }
+};
+
+//this plays the samples on the samplers with a click on the samplers buttons and stores its value for the sample vol knob to manipulate it
+DDJERGO.samplers = function(channel, control, value, status, group) {
+    const deck = status-0x90;
+    if (value === 0x7F) {
+        if (engine.getParameter(group, "play") === 1) {
+            engine.setParameter(group, "cue_default", 1);
+        } else {
+            engine.setParameter(group, "play", 1);
+        }
+        //store sample numbers
+        selectedSamples[0] = (control - 0x3C)/2 + (deck)*4;
+        selectedSamples[1] = (control - 0x3C)/2 + (deck)*4;
+    }
+};
+
+//this modifies the selected sample volume with sample vol knob rotations
+DDJERGO.samplerVolume = function(channel, control, value, status) {
+    const deck = status-0xB0;
+    let newValue;
+    if (value-64 > 0) { newValue = value-128; } else { newValue = value; }
+    if (newValue > 0) {
+        script.triggerControl(samplerStrings[selectedSamples[deck]], "pregain_up");
+    }
+    if (newValue < 0) {
+        script.triggerControl(samplerStrings[selectedSamples[deck]], "pregain_down");
+    }
+};
+
+//this plays the selected sample with a sample vol knob click
+DDJERGO.samplerKnobClick = function(channel, control, value, status) {
+    const deck = status-0x90;
+    const sample = selectedSamples[deck];
+    if (value === 0x7F) {
+        if (engine.getParameter(samplerStrings[sample], "play") === 1) {
+            engine.setParameter(samplerStrings[sample], "cue_default", 1);
+        } else {
+            engine.setParameter(samplerStrings[sample], "play", 1);
+        }
+    }
+};
+
+//this changes the sample stored in the knob being shift + turned
+DDJERGO.rotateSamples = function(channel, control, value, status) {
+    const deck = status-0xB0;
+    let newValue;
+    if (value-64 > 0) { newValue = value-128; } else { newValue = value; }
+    selectedSamples[deck] += newValue;
+    //makes sure samples stay within range
+    if (selectedSamples[deck] < 0) {
+        selectedSamples[deck] += 8;
+    }
+    if (selectedSamples[deck] > 7) {
+        selectedSamples[deck] -= 8;
+    }
+};
```

Comment:
Why 8? Don't you want each Sampler  Vol knob to control the samples of the respective side (1-4, 5-8)
Might be useful to know which sample is selected, e.g. turn the respective LED off for .5 sec? (off if a sample is loaded, on if not)

Reply:
I agree that some visual feedback is needed. However with the limitations of the controller I find it a bit complicated to provide clear and understandable feedback for the user. I remember using the FLX4 a few days ago and the effect selector drew a cursor over the selected effect. Is it possible somehow to highlight the selected sample on the screen? That'd be my first choice